src/python/pants/backend/helm/util_rules/chart.py

```diff
@@ -0,0 +1,253 @@
+
+    @classmethod
+    def from_dict(cls, d: dict[str, Any]) -> HelmChartMetadata:
+        deps = d.get("dependencies", [])
+
+        chart_type: ChartType | None = None
+        type_str = d.get("type")
+        if type_str:
+            try:
+                chart_type = ChartType(type_str)
+            except KeyError:
+                raise InvalidChartTypeValueError(type_str)
+
+        return cls(
+            api_version=d.get("apiVersion", _DEFAULT_API_VERSION),
+            name=d["name"],
+            version=d["version"],
+            icon=d.get("icon"),
+            app_version=d.get("appVersion"),
+            kube_version=d.get("kubeVersion"),
+            description=d.get("description"),
+            dependencies=tuple([HelmChartDependency.from_dict(dep) for dep in deps]),
+            type=chart_type or ChartType.APPLICATION,
+        )
+
+    @classmethod
+    def from_bytes(cls, content: bytes) -> HelmChartMetadata:
+        return cls.from_dict(yaml.safe_load(content))
+
+    def to_dict(self) -> dict[str, Any]:
+        d: dict[str, Any] = {
+            "apiVersion": self.api_version,
+            "name": self.name,
+            "version": self.version,
+            "type": self.type.value,
+        }
+        if self.icon:
+            d["icon"] = self.icon
+        if self.description:
+            d["description"] = self.description
+        if self.app_version:
+            d["appVersion"] = self.app_version
+        if self.kube_version:
+            d["kubeVersion"] = self.kube_version
+        if self.dependencies:
+            d["dependencies"] = list(self.dependencies)
+        return d
+
+    def to_yaml(self) -> str:
+        return cast("str", yaml.dump(self.to_dict()))
+
+
+@dataclass(frozen=True)
+class HelmChart:
+    address: Address
+    metadata: HelmChartMetadata
+    snapshot: Snapshot
+
+    lint_strict: bool | None
+
+    @property
+    def path(self) -> str:
+        return self.metadata.name
+
+
+_HELM_CHART_METADATA_FILENAMES = ["Chart.yaml", "Chart.yml"]
+
+
+def _chart_metadata_subset(digest: Digest) -> DigestSubset:
+    globs = PathGlobs(
+        [f"**/{filename}" for filename in _HELM_CHART_METADATA_FILENAMES],
+        glob_match_error_behavior=GlobMatchErrorBehavior.error,
+        conjunction=GlobExpansionConjunction.any_match,
+        description_of_origin="parse_chart_metadata",
+    )
+    return DigestSubset(digest, globs)
+
+
+@rule
+async def parse_chart_metadata_from_field(field: HelmChartMetaSourceField) -> HelmChartMetadata:
+    source_files = await Get(
+        HydratedSources,
+        HydrateSourcesRequest(
+            field, for_sources_types=(HelmChartMetaSourceField,), enable_codegen=True
+        ),
+    )
+    subset = await Get(Digest, DigestSubset, _chart_metadata_subset(source_files.snapshot.digest))
+    file_contents = await Get(DigestContents, Digest, subset)
+    return HelmChartMetadata.from_bytes(file_contents[0].content)
```

Comment:
Maybe `assert len(file_contents) == 1` or raise a more descriptive exception? Could there be multiple matches on chart metadata filenames? Probably not if digest is limited to just source files of this chart, but good to assert that the `[0]` access is valid.